### PR TITLE
asm: Add sx extension support.

### PR DESCRIPTION
**Pull request by rlavarec-gpsw**
docs: gcc https://gcc.gnu.org/onlinedocs/gcc/Overall-Options.html
      scons: https://scons.org/doc/3.0.5/HTML/scons-user/apb.html

-  .S      Windows:  assembly language file
          ARM: CodeSourcery Sourcery Lite
-  .sx     assembly language file + C pre-processor
          POSIX:  assembly language file + C pre-processor

Context:
 `file.S` added to the source to be precompiled under Windows.
Language c, cpp

`.S` is not found inside `can_compile`

meson throws an exception:
`meson.build:728:0: ERROR: No build machine compiler for 'src/file.S'`

From the file build.py
language c and cpp found, the '.S' is known inside `is_known_suffix `as `assembler_suffixes`, and throw
```
            val = input("Enter debug value: ")
            for s in sources:
                for lang, compiler in self.all_compilers.items():
                    if compiler.can_compile(s):
                        if lang not in self.compilers:
                            self.compilers[lang] = compiler
                        break
                else:
                    if is_known_suffix(s):
                        path = pathlib.Path(str(s)).as_posix()
                        m = f'No {self.for_machine.get_lower_case_name()} test-- machine compiler for {path!r}'
                        raise MesonException(m)
```

### CODE REVIEW

**Comment by rlavarec-gpsw** [comment]
Hello,
Sorry, I need to complete the context.
The need is to use cross compiling for android with 

cross config: [host_machine]
system = 'android'
cpu = 'arm64-v8a'
cpu_family = 'aarch64'
endian = 'little'

```
[binaries]
ar = 'c:/code/android_commandlinetools/9123335/ndk/25.1.8937393/toolchains/llvm/prebuilt/windows-x86_64/bin/llvm-ar.exe'
as = 'c:/code/android_commandlinetools/9123335/ndk/25.1.8937393/toolchains/llvm/prebuilt/windows-x86_64/bin/aarch64-linux-android-as'
nm = 'c:/code/android_commandlinetools/9123335/ndk/25.1.8937393/toolchains/llvm/prebuilt/windows-x86_64/bin/bin/llvm-nm'
ranlib = 'c:/code/android_commandlinetools/9123335/ndk/25.1.8937393/toolchains/llvm/prebuilt/windows-x86_64/bin/llvm-ranlib.exe'
strip = 'c:/code/android_commandlinetools/9123335/ndk/25.1.8937393/toolchains/llvm/prebuilt/windows-x86_64/bin/llvm-strip.exe'
pkgconfig = 'c:/code/stupeflix/Scripts/pkg-config.EXE'
c = 'aarch64-linux-android28-clang.cmd'
cpp = 'c:/code/android_commandlinetools/9123335/ndk/25.1.8937393/toolchains/llvm/prebuilt/windows-x86_64/bin/aarch64-linux-android28-clang++.cmd'
objc = 'aarch64-linux-android28-clang.cmd'
objcpp = 'c:/code/android_commandlinetools/9123335/ndk/25.1.8937393/toolchains/llvm/prebuilt/windows-x86_64/bin/aarch64-linux-android28-clang++.cmd'

[built-in options]
c_args = '-fPIC'
objc_args = '-fPIC'
c_link_args = '-fPIC'

[properties]
pkg_config_libdir = 'C:\code\devel\.sx\androidframeworks\arm64-v8a\lib\pkgconfig'
```
In case of cross

**Comment by rlavarec-gpsw** [comment]
Hello thanks,
I just give my debug output, we do use constant in python scripts.

`aarch64-linux-android28-clang.cmd `is given by the google sdk android command line tools




compiler stderr:
Microsoft (R) C/C++ Optimizing Compiler Version 19.35.32215 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.


Sanity testing C++ compiler: cl
Is cross compiler: False.
Sanity check compiler command line: cl sanitycheckcpp.cc /Fesanitycheckcpp.exe /MD /nologo /showIncludes /utf-8 /Zc:__cplusplus /link
Sanity check compile stdout:
sanitycheckcpp.cc

-----
Sanity check compile stderr:
```

-----
Running test binary command:  C:\code\build-android\arm64-v8a\nodegl\libnodegl\builddir-bundle\meson-private\sanitycheckcpp.exe
C++ compiler for the build machine: cl (msvc 19.35.32215 "Microsoft (R) C/C++ Optimizing Compiler Version 19.35.32215 for x64")
C++ linker for the build machine: link link 14.35.32215.0
detecting CPU family based on trial='amd64'
-----
Detecting compiler via: C:/code/android_commandlinetools/9123335/ndk/25.1.8937393/toolchains/llvm/prebuilt/windows-x86_64/bin/aarch64-linux-android28-clang++.cmd --version
compiler returned <Popen: returncode: 0 args: ['C:/code/android_commandlinetools/912...>
compiler stdout:
Android (8490178, based on r450784d) clang version 14.0.6 (https://android.googlesource.com/toolchain/llvm-project 4c603efb0cca074e9238af8b4106c30add4418f6)
Target: aarch64-unknown-linux-android28
Thread model: posix
InstalledDir: C:/code/android_commandlinetools/9123335/ndk/25.1.8937393/toolchains/llvm/prebuilt/windows-x86_64/bin

compiler stderr:

Running command: C:/code/android_commandlinetools/9123335/ndk/25.1.8937393/toolchains/llvm/prebuilt/windows-x86_64/bin/aarch64-linux-android28-clang++.cmd -E -dM -
Sanity testing C++ compiler: C:/code/android_commandlinetools/9123335/ndk/25.1.8937393/toolchains/llvm/prebuilt/windows-x86_64/bin/aarch64-linux-android28-clang++.cmd
Is cross compiler: True.
Sanity check compiler command line: C:/code/android_commandlinetools/9123335/ndk/25.1.8937393/toolchains/llvm/prebuilt/windows-x86_64/bin/aarch64-linux-android28-clang++.cmd sanitycheckcpp.cc -o sanitycheckcpp_cross.exe -D_FILE_OFFSET_BITS=64 -c
Sanity check compile stdout:

-----
Sanity check compile stderr:

-----
C++ compiler for the host machine: C:/code/android_commandlinetools/9123335/ndk/25.1.8937393/toolchains/llvm/prebuilt/windows-x86_64/bin/aarch64-linux-android28-clang++.cmd (clang 14.0.6 "Android (8490178, based on r450784d) clang version 14.0.6 (https://android.googlesource.com/toolchain/llvm-project 4c603efb0cca074e9238af8b4106c30add4418f6)")
C++ linker for the host machine: C:/code/android_commandlinetools/9123335/ndk/25.1.8937393/toolchains/llvm/prebuilt/windows-x86_64/bin/aarch64-linux-android28-clang++.cmd ld.lld 14.0.6
detecting CPU family based on trial='amd64'
Running compile:
Working directory:  C:\code\build-android\arm64-v8a\nodegl\libnodegl\builddir-bundle\meson-private\tmppt_gxjzr
Command line:  C:/code/android_commandlinetools/9123335/ndk/25.1.8937393/toolchains/llvm/prebuilt/windows-x86_64/bin/aarch64-linux-android28-clang++.cmd C:\code\build-android\arm64-v8a\nodegl\libnodegl\builddir-bundle\meson-private\tmppt_gxjzr\testfile.cpp -o C:\code\build-android\arm64-v8a\nodegl\libnodegl\builddir-bundle\meson-private\tmppt_gxjzr\output.obj -c -D_FILE_OFFSET_BITS=64 -O0 -fpermissive -Werror=implicit-function-declaration -Werror=unknown-warning-option -Werror=unused-command-line-argument -Werror=ignored-optimization-argument 

```

**Comment by rlavarec-gpsw** [comment]
ok, when the error throw:

`self.compilers.build['c'].id = 'msvc'`

```
[DEBUG] C compiler for the host machine: aarch64-linux-android28-clang.cmd (clang 14.0.6 "Android (8490178, based on r450784d) clang version 14.0.6 (https://android.googlesource.com/toolchain/llvm-project 4c603efb0cca074e9238af8b4106c30add4418f6)")
[DEBUG] C linker for the host machine: aarch64-linux-android28-clang.cmd ld.lld 14.0.6
[DEBUG] C compiler for the build machine: cl (msvc 19.35.32215 "Microsoft (R) C/C++ Optimizing Compiler Version 19.35.32215 for x64")
[DEBUG] C linker for the build machine: link link 14.35.32215.0
```

**Comment by xclaesse** [comment]
Do you build asm code for build machine? A target with `native: true`?

**Comment by rlavarec-gpsw** [comment]
There is no special treatment for the .S file

`lib_src += files('src/file.S')`

it is added to the `library('lib_name',lib_src)`
We use ninja, I try to debug to check where the choice of msvc is made, instead of clang.

**Comment by rlavarec-gpsw** [comment]
Ok I think I found the bug,
But I got no clue right now how to fix it.


At the point:
All is ok,


```
        logger_fun('Build machine cpu family:', mlog.bold(build_machine.cpu_family_method([], {})))
        logger_fun('Build machine cpu:', mlog.bold(build_machine.cpu_method([], {})))
        mlog.log('Host machine cpu family:', mlog.bold(host_machine.cpu_family_method([], {})))
        mlog.log('Host machine cpu:', mlog.bold(host_machine.cpu_method([], {})))
        logger_fun('Target machine cpu family:', mlog.bold(target_machine.cpu_family_method([], {})))
        logger_fun('Target machine cpu:', mlog.bold(target_machine.cpu_method([], {})))

```

But in the run

```
        try:
            if self.options.profile:
                fname = os.path.join(self.build_dir, 'meson-private', 'profile-interpreter.log')
                profile.runctx('intr.run()', globals(), locals(), filename=fname)
            else:
                intr.run()
```

Instead of taking the HOST like te first time before, `run `function set it to BUILD
`for_machine == BUILD`

```

        kwargs['include_directories'] = self.extract_incdirs(kwargs)
        target = targetclass(name, self.subdir, self.subproject, for_machine, srcs, struct, objs,
                             self.environment, self.compilers[for_machine], kwargs)
```

And here we got the code I got before,

".S" is invalid, because it is listed as known extension and not inside msvc compilation support for c and cpp!

We are getting there!

**Comment by rlavarec-gpsw** [comment]
I edited the PR to add sx extension

**Comment by rlavarec-gpsw** [comment]
Thank you very much Xavier for the support and this great project!

File: mesonbuild/compilers/compilers.py [hunk 1]
```diff
@@ -55,7 +55,7 @@
 # This means we can't include .h headers here since they could be C, C++, ObjC, etc.
 # First suffix is the language's default.
 lang_suffixes = {
-    'c': ('c',),
+    'c': ('c'),
```
Comment (bruchar1) [inline mesonbuild/compilers/compilers.py:58]: I think this should be a tuple, not a string.

Comment (xclaesse) [reply mesonbuild/compilers/compilers.py:58]: indeed, that coma is intentional, python syntax weirdness.

Comment (rlavarec-gpsw) [reply mesonbuild/compilers/compilers.py]: Fixed

**Pull request merged**